core/sequencer/election.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sequencer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/keytransparency/core/domain"
+	"github.com/google/keytransparency/core/mutator"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/util/election"
+)
+
+var isMaster monitoring.Gauge
+
+// Controller runs elections and controls receiver registration for a single domain.
+type Controller struct {
+	d                 *domain.Domain
+	factory           election.Factory
+	maxMastershipHold time.Duration
+}
+
+// NewController creates a new controller.
+func NewController(ctx context.Context,
+	d *domain.Domain,
+	factory election.Factory,
+	mastershipTimeout time.Duration,
+	mf monitoring.MetricFactory,
+) *Controller {
+	once.Do(func() {
+		isMaster = mf.NewGauge(
+			"master",
+			"Set to 1 for domains this server is master of",
+			domainIDLabel)
+	})
+	return &Controller{
+		maxMastershipHold: mastershipTimeout,
+		d:                 d,
+		factory:           factory,
+	}
+}
+
+// RunWhileMaster waits for mastership before registering a receiver
+// for domain d. During the receiver callback, check for continued mastership.
+func (c *Controller) RunWhileMaster(ctx context.Context,
+	newReceiver func(context.Context, *domain.Domain, func() error) (mutator.Receiver, error)) error {
+
+	el, err := c.factory.NewElection(ctx, c.d.DomainID)
+	if err != nil {
+		return fmt.Errorf("failed to create election for %v: %v", c.d.DomainID, err)
+	}
+	defer func(ctx context.Context) {
+		if err := el.Close(ctx); err != nil {
+			glog.Warningf("%d: Election.Close(): %v", c.d.DomainID, err)
+		}
+	}(ctx)
+	for {
+		if err := c.RunOnce(ctx, el, newReceiver); err != nil {
+			return err
+		}
+	}
+}
+
+// RunOnce waits for mastership, creates a new receiver, and cancels mastership after a bit.
+func (c *Controller) RunOnce(ctx context.Context, el election.MasterElection,
+	newReceiver func(context.Context, *domain.Domain, func() error) (mutator.Receiver, error)) error {
+	defer isMaster.Set(0, c.d.DomainID)
+	isMaster.Set(0, c.d.DomainID)
```

Comment:
I'm not sure it's redundant. 

1. The first line sets the variable to 0 while we're waiting to become master.  At this point we don't know if it's been previously set.
2. The second line sets the variable back to 0 when the function exists.

Reply:
> At this point we don't know if it's been previously set.

The only function that changes `isMaster` for a domain is `RunOnce`. Since it has a `defer isMaster.Set(0, c.d.DomainID`, it guarantees the invariant that `isMaster == 0` outside of `RunOnce` (even in presence of panics).